src/easylink/pipeline_graph.py

```diff
@@ -102,6 +83,92 @@ def merge_combined_implementations(self, config):
             except nx.NetworkXNoCycle:
                 pass
 
+    def get_combined_slots_and_edges(
+        self, combined_implementation: str, nodes_to_merge: list[str]
+    ) -> tuple[set[InputSlot], set[OutputSlot], set[EdgeParams]]:
+        slot_types = ["input_slot", "output_slot"]
+        combined_slots_by_type = combined_input_slots, combined_output_slots = set(), set()
+        edges_by_slot_and_type = self.get_edges_by_slot(nodes_to_merge)
+        transform_mappings = (InputSlotMapping, OutputSlotMapping)
+
+        combined_edges = set()
+
+        for slot_type, combined_slots, edges_by_slot, transform_mapping in zip(
+            slot_types, combined_slots_by_type, edges_by_slot_and_type, transform_mappings
+        ):
+            separate_slots = edges_by_slot.keys()
+            duplicate_slots = self.get_duplicate_slots(separate_slots, slot_type)
+            for slot_tuple in separate_slots:
+                step_name, slot = slot_tuple
+                edges = edges_by_slot[slot_tuple]
+                if slot_tuple in duplicate_slots:
+                    combined_slot = slot.__class__(
+                        name=step_name + "_" + slot.name,
+                        env_var=step_name.upper() + "_" + slot.env_var,
+                        validator=slot.validator,
+                    )
+                else:
+                    combined_slot = deepcopy(slot)
+                combined_slots.add(combined_slot)
+                slot_mapping = transform_mapping(
+                    slot.name, combined_implementation, combined_slot.name
+                )
+                combined_edges.update([slot_mapping.remap_edge(edge) for edge in edges])
+        return combined_input_slots, combined_output_slots, combined_edges
+
+    def get_edges_by_slot(
+        self,
+        nodes_to_merge: list[str],
+    ) -> tuple[
+        dict[tuple[str, InputSlot], EdgeParams], dict[tuple[str, OutputSlot], EdgeParams]
+    ]:
+
+        in_edges_by_slot = defaultdict(list)
+        out_edges_by_slot = defaultdict(list)
+
+        # Find separate input and output slots
+        for node in nodes_to_merge:
+            for pred, succ, data in self.in_edges(node, data=True):
+                if pred not in nodes_to_merge:
+                    input_slot = data["input_slot"]
+                    step_name = self.nodes[node]["implementation"].schema_steps[0]
```

Comment:
The `[0]` here seems a little strange. Plus, I don't think this would be disambiguated by e.g. loop or multiple. Any way this could use the step name directly from the node without going through the implementation?

Reply:
The assumption is that any node to be merged at this point only implements one step, so `schema_steps` is always a list of length 1. We could check that explicitly, if you think that's better
>  I don't think this would be disambiguated by e.g. loop or multiple
I think this shouldn't matter, at least for loops and multiples. For these, you should always be able to avoid merging two child nodes together, and even if you didn't, there isn't a relevant collision anyway, because the inputs are split from the same parent input slot and combined into the same outgoing edge. 
> Any way this could use the step name directly from the node without going through the implementation
I think reliably parsing the node name has more room for annoying exceptions than using an attribute does, but i didn't think too hard about it.

I think the only case that this might not cover is if you tried to combine an implementation that was already combined, but that configuration is precluded by validation I think

Reply:
> Surely all the nodes_to_merge will have the same implementation, right?
They do not, but the Implementations are essentially identical save for the fact that their `schema_steps` are different:
```
def get_implementation_graph(self) -> ImplementationGraph:
        """Return a single node with an implementation attribute."""
        implementation_graph = ImplementationGraph()
        combined_name = (
            self.pipeline_config[COMBINED_IMPLEMENTATION_KEY] if self.is_combined else None
        )
        implementation_node_name = self._step.implementation_node_name
        implementation = Implementation(
            schema_steps=[self._step.step_name],
            implementation_config=self.implementation_config,
            input_slots=self._step.input_slots.values(),
            output_slots=self._step.output_slots.values(),
            combined_name=combined_name,
        )
        implementation_graph.add_node_from_implementation(
            implementation_node_name,
            implementation=implementation,
        )
        return implementation_graph
```
> I could probably use a real-time conversation about this, maybe at mid-day?
Sure!